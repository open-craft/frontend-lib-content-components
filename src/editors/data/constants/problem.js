import { StrictDict } from '../../utils';

export const ProblemTypeKeys = StrictDict({
  TEXTINPUT: 'stringresponse',
  NUMERIC: 'numericalresponse',
  DROPDOWN: 'optionresponse',
  MULTISELECT: 'choiceresponse',
  SINGLESELECT: 'multiplechoiceresponse',
});

export const ProblemTypes = StrictDict({
  [ProblemTypeKeys.SINGLESELECT]: {
    title: 'Single Select Problem',
    preview: ('<div />'),
    description: 'Specify one correct answer from a list of possible options',
    helpLink: 'something.com',
  },
  [ProblemTypeKeys.MULTISELECT]: {
    title: 'Multi Select Problem',
    preview: ('<div />'),
    description: 'Specify one or more correct answers from a list of possible options.',
    helpLink: 'something.com',
  },
  [ProblemTypeKeys.DROPDOWN]: {
    title: 'Dropdown Problem',
    preview: ('<div />'),
    description: 'Specify one correct answer from a list of possible options, selected in a dropdown menu.',
    helpLink: 'something.com',
  },
  [ProblemTypeKeys.NUMERIC]: {
    title: 'Numeric Response Problem',
    preview: ('<div />'),
    description: 'Specify one or more correct numeric answers, submitted in a response field.',
    helpLink: 'something.com',
  },
  [ProblemTypeKeys.TEXTINPUT]: {
    title: 'Text Input Problem',
    preview: ('<div />'),
    description: 'Specify one or more correct text answers, including numbers and special characters, submitted in a response field.',
    helpLink: 'something.com',
  },
});

export const ShowAnswerTypesKeys = StrictDict({
  ALWAYS: 'always',
  ANSWERED: 'answered',
  ATTEMPTED: 'attempted',
  CLOSED: 'closed',
  FINISHED: 'finished',
  CORRECT_OR_PAST_DUE: 'correct_or_past_due',
  PAST_DUE: 'past_due',
  NEVER: 'never',
  AFTER_SOME_NUMBER_OF_ATTEMPTS: 'after_attempts',
  AFTER_ALL_ATTEMPTS: 'after_all_attempts',
  AFTER_ALL_ATTEMPTS_OR_CORRECT: 'after_all_attempts_or_correct',
  ATTEMPTED_NO_PAST_DUE: 'attempted_no_past_due',
});

export const ShowAnswerTypes = StrictDict({
  [ShowAnswerTypesKeys.ALWAYS]: {
    id: 'authoring.problemeditor.settings.showanswertype.always',
    defaultMessage: 'Always',
  },
  [ShowAnswerTypesKeys.ANSWERED]: {
    id: 'authoring.problemeditor.settings.showanswertype.answered',
    defaultMessage: 'Answered',
  },
  [ShowAnswerTypesKeys.ATTEMPTED]: {
    id: 'authoring.problemeditor.settings.showanswertype.attempted',
    defaultMessage: 'Attempted or Past Due',
  },
  [ShowAnswerTypesKeys.CLOSED]: {
    id: 'authoring.problemeditor.settings.showanswertype.closed',
    defaultMessage: 'Closed',
  },
  [ShowAnswerTypesKeys.FINISHED]: {
    id: 'authoring.problemeditor.settings.showanswertype.finished',
    defaultMessage: 'Finished',
  },
  [ShowAnswerTypesKeys.CORRECT_OR_PAST_DUE]: {
    id: 'authoring.problemeditor.settings.showanswertype.correct_or_past_due',
    defaultMessage: 'Correct or Past Due',
  },
  [ShowAnswerTypesKeys.PAST_DUE]: {
    id: 'authoring.problemeditor.settings.showanswertype.past_due',
    defaultMessage: 'Past Due',
  },
  [ShowAnswerTypesKeys.NEVER]: {
    id: 'authoring.problemeditor.settings.showanswertype.never',
    defaultMessage: 'Never',
  },
  [ShowAnswerTypesKeys.AFTER_SOME_NUMBER_OF_ATTEMPTS]: {
    id: 'authoring.problemeditor.settings.showanswertype.after_attempts',
    defaultMessage: 'After Some Number of Attempts',
  },
  [ShowAnswerTypesKeys.AFTER_ALL_ATTEMPTS]: {
    id: 'authoring.problemeditor.settings.showanswertype.after_all_attempts',
    defaultMessage: 'After All Attempts',
  },
  [ShowAnswerTypesKeys.AFTER_ALL_ATTEMPTS_OR_CORRECT]: {
    id: 'authoring.problemeditor.settings.showanswertype.after_all_attempts_or_correct',
    defaultMessage: 'After All Attempts or Correct',
  },
  [ShowAnswerTypesKeys.ATTEMPTED_NO_PAST_DUE]: {
    id: 'authoring.problemeditor.settings.showanswertype.attempted_no_past_due',
    defaultMessage: 'Attempted',
  },
});
