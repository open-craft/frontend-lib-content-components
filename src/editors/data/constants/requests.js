import { StrictDict } from '../../utils';

export const RequestStates = StrictDict({
  inactive: 'inactive',
  pending: 'pending',
  completed: 'completed',
  failed: 'failed',
});

export const RequestKeys = StrictDict({
  fetchAssets: 'fetchAssets',
  fetchBlock: 'fetchBlock',
  fetchImages: 'fetchImages',
  fetchUnit: 'fetchUnit',
  fetchStudioView: 'fetchStudioView',
  saveBlock: 'saveBlock',
  uploadAsset: 'uploadAsset',
  allowThumbnailUpload: 'allowThumbnailUpload',
  uploadThumbnail: 'uploadThumbnail',
  uploadTranscript: 'uploadTranscript',
  deleteTranscript: 'deleteTranscript',
  fetchCourseDetails: 'fetchCourseDetails',
  updateTranscriptLanguage: 'updateTranscriptLanguage',
  getTranscriptFile: 'getTranscriptFile',
});
