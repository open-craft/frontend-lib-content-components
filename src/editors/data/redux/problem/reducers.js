import { createSlice } from '@reduxjs/toolkit';
import { StrictDict } from '../../../utils';

const initialState = {
  rawOLX: '',
  problemType: 'SINGLESELECT',
  question: '',
  answers: [],
  groupFeedbackList: [],
  settings:{
    scoring: {
      advanced: false,
      scoring: {
        wieght: 0,
        attempts: {
          unlimited: true,
          number: 0,
        },
      },
    },
    hints: [],
    randomization: '',
    timeBetween: 0,
    MatLabApiKey: '',
    showAnswer: {
      on: '', // one of [OnAnswered, OnDueDate, AfterDueDate]
      afterAtempts: 1,
    },
    showResetButton: false,
  },
};

// eslint-disable-next-line no-unused-vars
const problem = createSlice({
  name: 'problem',
  initialState,
  reducers: {
    updateField: (state, { payload }) => ({
      ...state,
      ...payload,
    }),
    updateQuestion: (state, { payload }) => ({
      ...state,
      question: payload
    }),
    load: (state, { payload }) => ({
      ...state,
      ...payload,
    }),
    onSelect: (state, { payload }) => ({
      ...state,
      ...payload,
    }),
  },
});

const actions = StrictDict(problem.actions);

const { reducer } = problem;

export {
  actions,
  initialState,
  reducer,
};
