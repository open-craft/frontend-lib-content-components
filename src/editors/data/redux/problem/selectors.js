import { createSelector } from 'reselect';
import * as module from './selectors';


export const problemState = (state) => state.problem;
const mkSimpleSelector = (cb) => createSelector([module.problemState], cb);
export const simpleSelectors = {
  problemType: mkSimpleSelector(problemData => problemData.problemType),
  answers: mkSimpleSelector(problemData => problemData.answers),
  settings: mkSimpleSelector(problemData => problemData.settings),
  question: mkSimpleSelector(problemData => problemData.question),
}

export default {
  ...simpleSelectors,
};
