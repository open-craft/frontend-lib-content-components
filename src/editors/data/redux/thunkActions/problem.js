import { isEmpty, get } from 'lodash-es';
import { actions } from '..';
import { parseMarkdown } from '../../../containers/ProblemEditor/data/MarkDownParser';

export const initializeProblem = (blockValue) => (dispatch) => {
  const rawOLX = get(blockValue, 'data.data', {});
  const parsedData = get(blockValue, 'data.metadata.markdown', {});
  if (!isEmpty(parsedData) && !isEmpty(rawOLX)) {
    dispatch(actions.problem.load({ ...parseMarkdown(parsedData), rawOLX }));
  }
};

export default { initializeProblem };
