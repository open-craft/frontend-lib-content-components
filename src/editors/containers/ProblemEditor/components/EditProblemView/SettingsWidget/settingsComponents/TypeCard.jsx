import React from 'react'
import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';
import SettingsOption from '../SettingsOption';
import { Container, Icon } from '@edx/paragon';
import PropTypes from 'prop-types';
import { ProblemTypeKeys, ProblemTypes } from '../../../../../../data/constants/problem';
import { Check } from '@edx/paragon/icons';
import messages from '../messages';


const TypeRow = ({ name, selected, onClick, lastRow }) => (
    <>
        <Container size="xl" onClick={onClick} role='button' className="d-flex" fluid>
            <span className="flex-grow-1">{name}</span>
            <span hidden={selected}><Icon src={Check} className="text-success" /></span>
        </Container>
        <hr className={lastRow ? 'd-none' : 'd-block'} />
    </>
)


export const TypeCard = ({
    problemType,
    updateField,
    intl,
}) => {
    const problemTypeKeysArray = Object.values(ProblemTypeKeys);

    return (
        <SettingsOption
            title={intl.formatMessage(messages.typeSettingTitle)}
            summary={ProblemTypes[problemType].title}
        >
            {problemTypeKeysArray.map((typeKey, i) => (
                <TypeRow
                    key={i}
                    name={ProblemTypes[typeKey].title}
                    onClick={() => updateField({ problemType: typeKey })}
                    selected={typeKey !== problemType}
                    lastRow={(i + 1) == problemTypeKeysArray.length}
                />
            ))}
        </SettingsOption>
    )
}

TypeCard.propTypes = {
    hints: ProblemTypeKeys.isRequired,
    updateField: PropTypes.func.isRequired,
    intl: intlShape.isRequired,
};

export default injectIntl(TypeCard);