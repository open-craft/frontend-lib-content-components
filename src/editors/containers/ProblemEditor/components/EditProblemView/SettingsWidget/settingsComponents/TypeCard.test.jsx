import React from 'react';
import { shallow } from 'enzyme';
import { formatMessage } from '../../../../../../../testUtils';
import { TypeCard } from './TypeCard';
import { ProblemTypeKeys } from '../../../../../../data/constants/problem';

describe('TypeCard', () => {
  const props = {
    problemType: ProblemTypeKeys.TEXTINPUT,
    updateField: jest.fn().mockName('args.updateField'),
    intl: { formatMessage },
  };

  describe('snapshot', () => {
    test('snapshot: renders type setting card', () => {
      expect(shallow(<TypeCard {...props} />)).toMatchSnapshot();
    });
  });
});
