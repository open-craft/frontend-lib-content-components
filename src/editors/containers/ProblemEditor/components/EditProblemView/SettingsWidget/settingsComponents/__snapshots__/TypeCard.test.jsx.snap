// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeCard snapshot snapshot: renders type setting card 1`] = `
<SettingsOption
  summary="Text Input Problem"
  title="Type"
>
  <TypeRow
    key="0"
    label="Text Input Problem"
    lastRow={false}
    selected={false}
    typeKey="TEXTINPUT"
  />
  <TypeRow
    key="1"
    label="Numeric Response Problem"
    lastRow={false}
    selected={true}
    typeKey="NUMERIC"
  />
  <TypeRow
    key="2"
    label="Dropdown Problem"
    lastRow={false}
    selected={true}
    typeKey="DROPDOWN"
  />
  <TypeRow
    key="3"
    label="Multi Select Problem"
    lastRow={false}
    selected={true}
    typeKey="MULTISELECT"
  />
  <TypeRow
    key="4"
    label="Single Select Problem"
    lastRow={true}
    selected={true}
    typeKey="SINGLESELECT"
  />
</SettingsOption>
`;
